google/cloud/storage/internal/tracing_object_read_source.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/storage/internal/tracing_object_read_source.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/status.h"
+#include <opentelemetry/trace/scope.h>
+#include <chrono>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+TracingObjectReadSource::TracingObjectReadSource(
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
+    std::unique_ptr<storage::internal::ObjectReadSource> child)
+    : span_(std::move(span)), child_(std::move(child)) {}
+
+TracingObjectReadSource::~TracingObjectReadSource() {
+  // A download may be abandoned by the application, or kept open after all
+  // the data is received. Note that if there was an unrecoverable error the
+  // span is already finished in `TracingObjectReadOSource::Read()`.
+  internal::EndSpan(*span_);
+}
+
+bool TracingObjectReadSource::IsOpen() const {
+  auto scope = opentelemetry::trace::Scope(span_);
```

Comment:
comment: this is probably good practice in case we add any instrumentation to lower layers. But just an FYI: that creating a scope does a heap allocation, so it is relatively expensive.

Reply:
Is that true even of Noop spans?

Reply:
> Is that true even of Noop spans?

Yes. They do not check whether it is a valid span before pushing to the context stack (and creating a token to track it).

https://github.com/open-telemetry/opentelemetry-cpp/blob/6dbfdb52ebca82d8705392fe88dadd9c133374e4/api/include/opentelemetry/trace/scope.h#L32-L35